bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -164,248 +214,330 @@ enum RequestDisposition {
+    /// popped first (even if it arrived later in time than the next component
+    /// change request).
     pub fn pop_front(&mut self) -> Option<ExternalRequest> {
-        self.queue.pop_front()
+        if let Some(state_change) = self.state_queue.pop_front() {
+            Some(ExternalRequest::State(state_change))
+        } else {
+            self.component_queue.pop_front().map(ExternalRequest::Component)
+        }
     }
 
     /// Indicates whether the queue is empty.
     #[cfg(test)]
     pub fn is_empty(&self) -> bool {
-        self.queue.is_empty()
+        self.state_queue.is_empty() && self.component_queue.is_empty()
     }
 
-    /// Asks to place the supplied request on the queue. If the request is
-    /// enqueued, updates the dispositions to use for future requests.
+    /// Attempts to replace the supplied `request` on the queue, returning `Ok`
+    /// if the request was accepted and an `Err` otherwise. In the latter case,
+    /// the error contains a [`RequestDeniedReason`] that describes why the
+    /// request was rejected.
     pub fn try_queue(
         &mut self,
         request: ExternalRequest,
     ) -> Result<(), RequestDeniedReason> {
-        let disposition = match request {
-            ExternalRequest::Start => self.allowed.start,
-            ExternalRequest::MigrateAsSource { .. } => {
-                self.allowed.migrate_as_source
-            }
-            ExternalRequest::Reboot => self.allowed.reboot,
-            ExternalRequest::ReconfigureCrucibleVolume { .. } => {
-                self.allowed.mutate
+        // If the queue is in a terminal state, deny the request straightaway
+        // (unless it's a stop request, which can be ignored for idempotency).
+        let disposition = if let Some(reason) = self.state.deny_reason() {
+            if matches!(
+                request,
+                ExternalRequest::State(StateChangeRequest::Stop)
+            ) {
+                Disposition::Ignore
+            } else {
+                Disposition::Deny(reason)
             }
+        } else {
+            // The instance hasn't stopped yet, so consider this request in
+            // light of its current state and the other as-yet unprocessed
+            // requests from the queue.
+            //
+            // In general, try to make state change requests idempotent by
+            // ignoring new requests when a request of the appropriate kind is
+            // already on the queue, and deny requests to reach a state that is
+            // precluded by an earlier state change request.
+            match request {
+                // Interpret start requests as requests to reach the Running
+                // state.
+                ExternalRequest::State(StateChangeRequest::Start) => {
+                    if self.awaiting_stop {
+                        Disposition::Deny(RequestDeniedReason::HaltPending)
```

Comment:
part of me wonders if the `awaiting_stop` logic common to every state change request except `stop` could be factored out, but...idk if it's worth it; i think expressing the complete logic for each possible state request in the match arm for that request is probably clearer than trying to be less repetitive...